xdsl/parser.py

```diff
@@ -858,6 +876,66 @@ def _parse_dialect_type_or_attribute_inner(self, kind: str) -> Attribute:
             return cast(Data[Any], type_def(param))
         assert False, "Attributes are either ParametrizedAttribute or Data."
 
+    def _parse_unregistered_attr_body(self) -> str:
+        """
+        Parse the body of an unregistered attribute, which is a balanced
+        string for `<`, `(`, `[`, `{`, and may contain string literals.
+        """
+        self._synchronize_lexer_and_tokenizer()
+        self._parse_token(Token.Kind.LESS, "Expected '<' for attribute body!")
+
+        start_pos = self._current_token.span.start
+        end_pos: int = start_pos
+
+        symbols_stack = [Token.Kind.LESS]
+        parentheses = {
+            Token.Kind.GREATER: Token.Kind.LESS,
+            Token.Kind.R_PAREN: Token.Kind.L_PAREN,
+            Token.Kind.R_SQUARE: Token.Kind.L_SQUARE,
+            Token.Kind.R_BRACE: Token.Kind.L_BRACE
+        }
+        parentheses_names = {
+            Token.Kind.GREATER: '`>`',
+            Token.Kind.R_PAREN: '`)`',
+            Token.Kind.R_SQUARE: '`]`',
+            Token.Kind.R_BRACE: '`}`'
+        }
+        while True:
+            # Opening a new parenthesis
+            if (token := self._parse_optional_token_in(
+                    parentheses.values())) is not None:
+                symbols_stack.append(token.kind)
+                continue
+
+            # Closing a parenthesis
+            if (token := self._parse_optional_token_in(
+                    parentheses.keys())) is not None:
+                closing = parentheses[token.kind]
+                if symbols_stack[-1] != closing:
+                    self.raise_error(
+                        "Mismatched {} in attribute body!".format(
+                            parentheses_names[self._current_token.kind]),
+                        self._current_token.span)
+                symbols_stack.pop()
+                if len(symbols_stack) == 0:
+                    break
+                end_pos = self._current_token.span.start
+                continue
```

Comment:
This handles `->` because our tokenizer is smart, right?

Reply:
Yes!